8333658: NMT: Use an allocator with 4-byte pointers to save memory in NativeCallStackStorage

jdksjolen opened the pull request:
Hi,

This PR introduces a new allocator, `IndexedFreelistAllocator`. It uses a `GrowableArray` in order to have 4-byte "pointers" to its elements and also works as a freelist as unused slots form a linked list. This allocator cannot shrink its used memory. I'm always open for better names.

We then use this allocator in order to store the `NativeCallStackStorage` hash table. This saves `4 * 4099 = ~16KiB` of memory as each table element is now only 4 bytes. For each stored stack it also saves 4 bytes. The fact that the allocator cannot shrink is fine, we do not ever remove stacks from the storage.

The main point of this is to introduce a pointer-generic experimential API, so I also implemented CHeap and Arena allocators. It's currently placed in NMT, but we might want to move it into utilities. It uses a bit of template-magic, but my IDE (Emacs+clangd) was actually able to figure things out when the types didn't line up correctly etc, so it's not an enemy to IDE help.

It sounded expensive to have the GrowableArray continously realloc its underlying data array, so I did a basic test where we allocate 1 000 000 stacks and push them into NativeCallStackStorage backed by different allocators. This is available in the PR.

The results are as follows on linux-x64-slowdebug:

```
Generate stacks... Done
Time taken with GrowableArray: 8341.240945
Time taken with CHeap: 12189.031318
Time taken with Arena: 8800.703092
Time taken with GrowableArray again: 8295.508829
```

And on linux-x64:

```
Time taken with GrowableArray: 8378.018135
Time taken with CHeap: 12437.347868
Time taken with Arena: 8758.064717
Time taken with GrowableArray again: 8391.076291
```

Obviously, this is a very basic benchmark, but it seems like we're faster than CHeap and Arena for this case.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333658](https://bugs.openjdk.org/browse/JDK-8333658): NMT: Use an allocator with 4-byte pointers to save memory in NativeCallStackStorage (**Enhancement** - P4)


### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [3d9e324c](https://git.openjdk.org/jdk/pull/18979/files/3d9e324c31007909d2e67ab16e40f1877a395476)
 * [Afshin Zafari](https://openjdk.org/census#azafari) (@afshin-zafari - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18979/head:pull/18979` \
`$ git checkout pull/18979`

Update a local copy of the PR: \
`$ git checkout pull/18979` \
`$ git pull https://git.openjdk.org/jdk.git pull/18979/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18979`

View PR using the GUI difftool: \
`$ git pr show -t 18979`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18979.diff">https://git.openjdk.org/jdk/pull/18979.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18979#issuecomment-2150468082)

jdksjolen commented:
An obvious improvement: Have the returned pointers remember which allocator object it was allocated from when in debug mode to avoid free:ing an element using the wrong allocator. You can also support detecting double frees by assigning each allocation a unique id.

tstuefe commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+    } else {
+      // Follow the link to the next free element
+      free_start = be.link;
+    }
```
This feels a bit uncommon and contrary to the normal linked list implementation. If possible, I would opt for a more traditional approach that is easier to understand and does not rely on at_grow (e.g. in case we want adapt this to some other form of backing storage).

e.g.

```
if (_freelist == -1) {
  allocate new slot   
} else {
  reuse slot at freelist
}
```

tstuefe replied:
Note that at_grow also could have the problem that an error, resulting in a high index, would cause a large allocation to happen.

I think this class could use an optional max value.

tstuefe commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+    BackingElement& be_freed = backing_storage.at(i.idx);
+    be_freed.link = free_start;
+    free_start = i;
+  }
```
I don't think a free(index) is very useful. We need a free(E* e), that calculates I from &e. (Your one real world usage example, NativeCallStackStorage, conveniently never frees :=)

tstuefe replied:
And we may want a function, complementing `at()`, that gives us the index for a given element pointer.

jdksjolen replied:
I did start a port of the Treap to this allocator interface and there was no particular problem with using `free(I)` instead of `free(E* e)`. I gave up on that because it was a much larger change and I didn't know what the response would be to the interface itself (plus, we really do want to give back memory). Performing an inversion might not be easy. Consider something like a chunk allocator like our arena, translating index (`uint16_t chunk; uint16_t index;`) to a pointer requires walking the chunks.

jdksjolen replied:
The tests also uses the freeing functionality.

jdksjolen replied:
>). Performing an inversion might not be easy. Consider something like a chunk allocator like our arena, translating index (uint16_t chunk; uint16_t index;) to a pointer requires walking the chunks.

tstuefe commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+    return backing_storage.at(i.idx).e;
+  }
+};
+
```
What are these allocators for?

jdksjolen replied:
So you can easily swap out your allocator if you have a bug and want to ensure the bug isn't in your allocator. Also just completeness, and necessary for the performance tests. We can remove them, but I think they're nice to have.

tstuefe reviewed (commented):
Started looking at this. I like this.

The smaller beauty spot of your implementation is that by relying on GrowableHeapArray, we are bound to the size of its index type, which is int. So we will never be able to hold more than 2G-1 entries. I think thats okay for now though, and if not, its easy to change in GrowableArray.

A bigger beauty spot is that it needs explicit initialization for every member (IIUC). That feels just wasteful, especially in combination with the generous hard-coded power-of-2 growth of GrowableArray.

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+    }
+    BackingElement(E& e) {
+      this->e = e;
+    }
```
That's a good idea. It probably should be:

```c++
union alignas(E) BackingElement  {
  I link;
  char e[sizeof(E)];
};
```

tstuefe replied:
Even better

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+  }
+
+  E& operator[](I i) {
+    return backing_storage.at(i.idx).e;
+  }
```
It's nice when your other operation is called `translate`, maybe not as nice when it's called `at`.

jdksjolen reviewed (commented):
Thanks Thomas! Sorry for the messy code, I felt that it was in good enough shape to show people and see what they think. I'll clean it up and ping you when it's ready. I answered some of your comments also.

/ Johan

jdksjolen commented:
I actually messed up the performance numbers for `linux-x64` and accidentally just ran the slow-debug test again :). Here are the actual results, now we're beating Arena and CHeap even harder than before so that's great.

```
Generate stacks... Done
Time taken with GrowableArray: 3521.723657
Time taken with CHeap: 8549.043628
Time taken with Arena: 5510.189990
Time taken with GrowableArray again: 3535.035375
```

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+    bool operator==(I other) {
+      return idx == other.idx;
+    }
+  };
```
Now we use access specifiers to deny access to the index, except for the allocator, which is a friend. We do want the index to be opaque, so we do this. On top of that, we use an `_owner` field in debug mode to assert that we don't free to the wrong allocator.

tstuefe replied:
> Now we use access specifiers to deny access to the index, except for the allocator, which is a friend. We do want the index to be opaque, so we do this. On top of that, we use an `_owner` field in debug mode to assert that we don't free to the wrong allocator.

The last part I don't like. It causes `I` to quadruple in size in debug builds only. For something that is potentially used as member in bunch of different data structures, it can cause considerable deltas in memory layout between debug and release builds. I don't think that is a good thing. What you test in debug should be close to what's running at a customer.

I also don't think it is particularly useful. I would instead do a simple range check on I. That gives you a part of the sanity checks without the negatives. Couple that with a sanity check wrt to slot state (don't free free slots) you are pretty well covered already.

Other than that, I think using an own class to hide a simple i32 is unnecessary complex, but I leave that up to you. If you do this, please take care of constness. And make it work for its money, e.g. by giving it an is_nil instead of exposing nil, and requiring the user to manually compare it.

jdksjolen replied:
Alright, I get the counterpoint re: memory usage of the slots. I can live with that, though I am a bit disappointed as I had double-free coverage in mind also :).

>Other than that, I think using an own class to hide a simple i32 is unnecessary complex, but I leave that up to you. If you do this, please take care of constness. And make it work for its money, e.g. by giving it an is_nil instead of exposing nil, and requiring the user to manually compare it.

Yes, I'll make `_idx` const. `nil` is also exposed so that you can create nil pointers, which is important. I wanted to make the `I(int32_t idx)` constructor private, as only the allocator should construct pointers. This turned out to not be possible due to the design of `GrowableArray`.

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+
+  IndexedFreeListAllocator()
+  : backing_storage(8),
+  free_start(I{0}) {}
```
>Also, please expose the init size to outside, possibly with a sensible default like your 8

Yeah, we can do this.

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,149 @@
+
+  E& translate(I i) {
+    return backing_storage.at(i.idx).e;
+  }
```
Out-of-bounds should be taken care of by GrowableArray, luckily.

tstuefe commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,198 @@
+    I link;
+    char e[sizeof(E)];
+
+    BackingElement() {
```
Init list?

jdksjolen replied:
It's on its way, I haven't pushed all of my changes yet :-).

jdksjolen reviewed (commented):
I've got some stuff left to do:

1. Clean up the tests
2. Fix the maximum bound that you suggested

Btw, I figured it out: Github shows my comments twice: Once as part of the review I'm making, and once separately where they should be. I make and submit reviews because that bunches my e-mails up in one long one for the mailing lists -- I think it does so at least.

jdksjolen commented:
I need to take a think about this, it's quite annoying that we're generating a new `I` per `IFLA<E, flag>`. I think that we could make it a global struct `IndexPointer`, or something like that.

tstuefe commented:
> I need to take a think about this, it's quite annoying that we're generating a new `I` per `IFLA<E, flag>`. I think that we could make it a global struct `IndexPointer`, or something like that.

TBH, I am quite opinionated about this. One example:

Yesterday, I tried to rework the VMATreeTest.TestConsistencyWithSimpleTracker_vm. See issue https://bugs.openjdk.org/browse/JDK-8334179. I first tried to implement a more efficient memory layout for the SimpleVMATracker (everything in SimpleVMATracker::Info could, essentially, stored in a single byte, giving us a much better cache performance). But for that, I would have needed the numerical value of the stack index. But that is hidden away in NativeCallStackStorage::StackIndex. Apart from giving me RSI from typing, I cannot just extract the integer from NativeCallStackStorage::StackIndex since its sole purpose in life is to hide that integer. Of course, I can either declare yet another friend class. But that's ugly and does not scale.

I really abhor the friend class concept since it subverts good encapsulation. It means I have different classes that can freely rummage around in my innards. So I need IDE introspection to actually know whats going on. For a terrible example of how complex this gets see Chunk and Arena and all its child classes, and try to figure out who does modify the Arena innards when and why. Compare that with Metaspace, which is the spiritual successor of Arenas. I am biased oc, but IMHO Metaspace has a much cleaner separation of concerns and encapsulation.

I can kind of see the point with test classes, but even then, a better way would be to make the test class an inner class.

TL;DR: I think these wrappers make life difficult while giving only minuscule protection against errors - if any at all.

tstuefe commented:
Some general remarks, (apart from my remark about the integer wrapper), and then I wait until you say its ready for review.

1) I would remove/not bother with the Allocators. There isn't much of a point, and small code is good code. The arena version is particularly questionable since Arenas don't really support arbitrary deallocation. I would not want anyone to use this allocator in real code.

2) I would like to have a non-growing version. Optionally, one where I can hand in an address range, and that gets used. Could possibly be combined for simplicity (if you specify a range, its a non-growing array). Reason: I want to be able to place stuff that needs to be address-stable, and I often need to do this in pre-allocated ranges.

jdksjolen commented:
>But for that, I would have needed the numerical value of the stack index. But that is hidden away in NativeCallStackStorage::StackIndex. Apart from giving me RSI from typing, I cannot just extract the integer from NativeCallStackStorage::StackIndex since its sole purpose in life is to hide that integer. Of course, I can declare yet another friend class. But that's ugly and does not scale.

Alright, I've crumbled. I changed it to `using I = int32_t;`. Feel free to do the same with `StackIndex`.

>I can kind of see the point with test classes, but even then, a better way would be to make the test class an inner class.

Noo, it's so painful to have test code embedded within prod code.

>I would remove/not bother with the Allocators. There isn't much of a point, and small code is good code. The arena version is particularly questionable since Arenas don't really support arbitrary deallocation. I would not want anyone to use this allocator in real code.

That's fine, I've removed the performance testing.

>I would like to have a non-growing version. Optionally, one where I can hand in an address range, and that gets used. Could possibly be combined for simplicity (if you specify a range, its a non-growing array). Reason: I want to be able to place stuff that needs to be address-stable, and I often need to do this in pre-allocated ranges.

I see the value in that, but can we do that as a separate PR? Is this a Metaspace thing, by the way?

**Note:** We still have no checks for double-frees. That'll need some baking in my head.

jdksjolen commented:
Also: Obviously, feel fine with removing `StackIndex` as a wrapper.

tstuefe commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -73,16 +54,16 @@ class IndexedFreeListAllocator {
   template<typename... Args>
   I allocate(Args... args) {
     BackingElement* be;
     int i;
```
I i? Then, later, just return i?

jdksjolen replied:
Thanks, clean up from wrapping I.

tstuefe commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -95,103 +76,24 @@ class IndexedFreeListAllocator {
-    assert(!i.is_nil() || (i._idx > 0 && i._idx < _backing_storage.length()), "out of bounds free");
-    if (i.is_nil()) return;
-    BackingElement& be_freed = _backing_storage.at(i._idx);
+    assert(i != nil || (i > 0 && i < _backing_storage.length()), "out of bounds free");
```
I think there are some errors here. This is probably broken. Which we would see if the gtests were running, but hotspot common tier1 tests seem broken.

Do we allow passing in nil? Then, i must be either nil or valid, not != nil or valid. If not, use an AND, not an OR.
i=0 is valid
Could you also please factor out OOB test for i?

jdksjolen replied:
It's meant to express `P => Q` equiv. to `~P V Q`, but we're trying to say "is_not_nil(i) => is_valid(i)`, so it should be `i == nil` **not** `i != nil`. Yes, it's broken. Also yes, let's make OOB check a predicate function.

tstuefe commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -95,103 +76,24 @@ class IndexedFreeListAllocator {
-    if (i.is_nil()) return;
-    BackingElement& be_freed = _backing_storage.at(i._idx);
+    assert(i != nil || (i > 0 && i < _backing_storage.length()), "out of bounds free");
+    if (i != nil) return;
```
i == nil?

jdksjolen replied:
Not sure how I managed this miracle of coding :), yes.

tstuefe reviewed (commented):
More comments. Thanks for taking my suggestions.

We need a little gtest for this.

jdksjolen commented:
@tstuefe,

Is there anything left in this that you'd like to see?

@gerard-ziemski, @afshin-zafari , may I receive a second review on this?

tstuefe commented on src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:
```diff
@@ -40,10 +40,11 @@
 // We achieve this by using a closed hashtable for finding previously existing NCS:s and referring to them by an index that's smaller than a pointer.
-class NativeCallStackStorage : public CHeapObj<mtNMT> {
+template<template<typename, MEMFLAGS> class ALLOCATOR>
+class NativeCallStackStorageWithAllocator : public CHeapObj<mtNMT> {
```
I don't think we need a templatized version of NativeCallStackStorage. Please let's keep it simple and stupid. We will only ever have one form of NativeCallStackStorage.

Should we ever need multiple forms of NativeCallStackStorage simultaneously, we can still templatize.

tstuefe replied:
Making this a template also makes it difficult to move implementation over to a cpp file. Something which you really should do, since the implementation here is non-trivial.

tstuefe commented on src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:
```diff
@@ -68,36 +69,46 @@ class NativeCallStackStorage : public CHeapObj<mtNMT> {
+    LinkPtr next;
     StackIndex stack;
-    Link(Link* next, StackIndex v)
+    Link(LinkPtr next, StackIndex v)
```
Pre-existing. We should never modify the hash table, so the stackindex can be const.

tstuefe reviewed (commented):
@jdksjolen sorry, had this review sitting in pending state for days appearantly and forgot to send it off.

I don't understand. I thought the point of this new class is to move the **stacks** over to the new allocator-with-freelist, and make StackIndex a simple typedef for the allocator index? Then, get rid of the GrowableArray?

You put the hashtable entries in there, which is okay, but not using the full potential of this new class.

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,102 @@
+template<typename E, MEMFLAGS flag>
+class IndexedFreeListAllocator {
+public:
+  using I = int32_t;
```
This is what I'd like to have in a follow-up RFE. Your last two points are because of `GrowableArray`: GA uses int32, so we do, and GA checks overflow when expanding, so we're OK.

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  const E& at(I i) const {
+    assert(i != nil, "null pointer dereference");
```
You're probably right, I'm not sure what a `const IFLA` would be.

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,102 @@
+    return i;
+  }
+
+  void free(I i) {
```
Just `deallocate`?

jdksjolen commented on src/hotspot/share/nmt/indexedFreeListAllocator.hpp:
```diff
@@ -0,0 +1,102 @@
+// It is "indexed" as a reference is base + index * sizeof(E).
+// It never returns any memory to the system.
+template<typename E, MEMFLAGS flag>
+class IndexedFreeListAllocator {
```
"HomogenousObjectArray"

jdksjolen commented:
>One thing I don't understand. I thought the point of this new class is to move the stacks over to the new allocator-with-freelist, and make StackIndex a simple typedef for the allocator index? Then, get rid of the GrowableArray?
>You put the hashtable entries in there, which is okay, but not using the full potential of this new class.

No, the initial point was to get rid of pointer-heavy structure's unnecessarily large pointers. The discussion of StackIndex and its wrapping came later in the review process. Moving the `StackIndex` to a simple typedef is its own RFE. If I were to use `HOA` for the stacks in this PR, then I wouldn't do the refactoring of `StackIndex` into a typedef :-).

tstuefe commented on src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:
```diff
@@ -68,36 +68,46 @@ class NativeCallStackStorage : public CHeapObj<mtNMT> {
+  using LinkPtr = typename Allocator::I;
+  LinkPtr nil() { return Allocator::nil; }
+
+  Allocator _allocator;
```
"Allocator" is misleading. An allocator is typically a thing that provides alloc() and free() for you, but does not maintain the storage for you, thats up to you. 

Proposal: Allocator -> TableEntryStorage or EntryStorage. _allocator = _entry_storage.

tstuefe commented on src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:
```diff
@@ -68,36 +68,46 @@ class NativeCallStackStorage : public CHeapObj<mtNMT> {
+  struct Link;
+  using Allocator = HomogenousObjectArray<Link, mtNMT>;
+  using LinkPtr = typename Allocator::I;
+  LinkPtr nil() { return Allocator::nil; }
```
could we have some clearer names?

- Link -> TableEntry or Entry (since its entries of the hashtable)
- LinkPtr -> TableEntryIndex

Also, could you put all data members of NCCS here, please? It makes it easier to see the implementation. The big boys here are _table, hash table entry storage, and stack storage.

I don't think you need a nil(), at least not if you keep referring to Allocator::nil elsewhere. 

Btw, why do you prefer using over  typedef?

jdksjolen replied:
All suggestions sounds fine 

> Btw, why do you prefer using over typedef?

It's helpful that the name sticks out a lot more than for typedef when doing stuff like function pointer or array aliases, the name is always left of `=`. `using` also supports templates (you can do `template <typename T> using x = array<T>`).

That's why you might prefer `using` to `typedef`. I prefer it because that's what I'm used to using, boring but true.

jdksjolen commented:
Surely a problem with `HomogenousObjectArray` is that all arrays store objects of the same type.

jdksjolen commented:
Hi,

I've changed the data which is stored in `HOA` to be trivial and have added `static_assert`s to the methods. We **cannot** add the `static_assert` to the top of the `HOA` class because there's a mutual dependency on the data type being stored and the `HOA`. It's very important that the data we put into `HOA` is trivial as we do not call any destructors or copy constructors/operators on the code. **Any such "improvements" may be undesirable**, it's good to have limited containers and data which is simpler to reason about.

Here's a minimal example show casing the problem of why `static_assert` has to live in the methods and not in the class definition:

```c++
#include <type_traits>
#include <unordered_map>

template<typename E>
struct A {
    // SEE ME: Uncomment this line to see compilation errors re:
    // "incomplete types"
    // static_assert(std::is_trivial<E>::value);
    union alignas(E) Foo {
        char e[sizeof(E)];
    };
    using I = int;

    Foo* foo;
    void put(E& e) {
        static_assert(std::is_trivial<E>::value);
        }
};

struct C {
    struct B;
    using AB = A<B>;
    using I = typename AB::I;

    struct B {
        I i;
        // SEE ME! Uncomment this line to make B non-trivial
        // std::unordered_map<int, int> s;
    };

    AB a;

    void do_it() {
        B b;
        a.put(b);
    }
};

int main() {
    C c;
    c.do_it();
    return 0;
}
``` 

I suggest that you paste this in godbolt.org (remember to switch to C++) or compile locally, following the `SEE ME` comments to understand what's going on.

afshin-zafari commented on src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:
```diff
@@ -26,8 +26,8 @@
 #include "memory/allocation.hpp"
-#include "memory/arena.hpp"
 #include "utilities/growableArray.hpp"
+#include "nmt/homogenousObjectArray.hpp"
```
order of headers.

jdksjolen replied:
Thanks! Fixed

afshin-zafari reviewed (commented):
Thanks Johan. One small nit on order of header files.

stefank commented:
Could we try another round of coming up with a better name for this utility? HomogenousObjectArray is eerily similar to G1's humongous object arrays. It's also not clear to me what makes this array an homogenous array. Is our other arrays non-homogenous?

Would it make sense to put this in utilities/ instead of nmt/.

The include guards should be placed before the includes.

jdksjolen commented:
>The include guards should be placed before the includes.

Thanks!

>Would it make sense to put this in utilities/ instead of nmt/.

Do we have anyone else that wants to use this? The threshold in terms of API quality is typically higher for something in utilities/ than a local utility. I'm working on some improvements to this allocator already, maybe we can take a move to utilities/ together with those improvements?


> Could we try another round of coming up with a better name for this utility? HomogenousObjectArray is eerily similar to G1's humongous object arrays. It's also not clear to me what makes this array an homogenous array. Is our other arrays non-homogenous?

I agree on the name part. We could call it a `HomogenousAllocator`, that does differentiate it meaningfully from something like `Arena` which can allocate anything. @tstuefe , arenas are also an example of the storage/lifetime of the objects from an allocator being bound to the lifetime of that allocator, so I think this makes sense.

tstuefe commented:
> > The include guards should be placed before the includes.
> 
> Thanks!
> 
> > Would it make sense to put this in utilities/ instead of nmt/.
> 
> Do we have anyone else that wants to use this? The threshold in terms of API quality is typically higher for something in utilities/ than a local utility. I'm working on some improvements to this allocator already, maybe we can take a move to utilities/ together with those improvements?

Yes, I will use it to replace some code in Metaspace at least, but for that I need some more features (placing this thing atop of an existing memory range, and templatized index type). 

Can all be done in a follow-up RFE.

> 
> > Could we try another round of coming up with a better name for this utility? HomogenousObjectArray is eerily similar to G1's humongous object arrays. It's also not clear to me what makes this array an homogenous array. Is our other arrays non-homogenous?
> 
> I agree on the name part. We could call it a `HomogenousAllocator`, that does differentiate it meaningfully from something like `Arena` which can allocate anything. @tstuefe , arenas are also an example of the storage/lifetime of the objects from an allocator being bound to the lifetime of that allocator, so I think this makes sense.

Names are hard.

We already have two types of arenas in hotspots, plus glibc has arenas, so "arena" is not ideal either. Arena does usually not imply a free list either, nor homogenous sizes.

Thinking about it, Array already sort of implies homogeneous sizes, so maybe "homogeneous" is redundant.

ArrayWithFreeList? Unsexy but precise.

tstuefe commented:
> Hi,
> 
> I've changed the data which is stored in `HOA` to be trivial and have added `static_assert`s to the methods. ... It's very important that the data we put into `HOA` is trivial as we do not call any destructors or copy constructors/operators on the code. 

We don't, but why forbid the user from doing that? I may want to place a non-trivial object in there, e.g. one where I forbid copying altogether, or one with a single non-default ctor? What would be the problem with that?

tstuefe commented on test/hotspot/gtest/nmt/test_homogenousObjectArray.cpp:
```diff
@@ -0,0 +1,152 @@
+  alloc.deallocate(i1);
+  A::I i3 = alloc.allocate(0);
+  EXPECT_EQ(p1, &alloc.at(i3));
+}
```
I think for what little the tests do, testing with different list types is overengineered. You could just scrap both LL and LL2, and just use a HOA directly.

tstuefe replied:
Hmm. I am not sold on the "example" benefit. The array is quite easy in itself.

You could scratch the lists, since their implementations don't do anything to really test the AWFL (new acronym, yay). For the saved LOCs could expand the tests to test with a collection of various types, e.g.
- u1, u2, u8
- unaligned structures that need alignment (e.g. struct (void*; int; ))
- trivial objects

Interesting to see would be that it works, that alignment is correct, that nothing breaks across resizes (indexes stay stable, etc.).

I wont insist of it, just thinking that the current test complexity is somewhat wasted.

jdksjolen replied:
>really test the AWFL (new acronym, yay).

Can't wait for templatized indices so I can write

```c++
using AWFUL = ArrayWithFreeListAllocator<int, uint8_t>;
```

>Interesting to see would be that it works, that alignment is correct, that nothing breaks across resizes (indexes stay stable, etc.).

>I wont insist of it, just thinking that the current test complexity is somewhat wasted.

I think those tests make sense if we replace the `GrowableArray`as the backing memory area with something else, otherwise that's just testing the GA.

OK, how about we scrap the tests for when moving this to utilities and having the index being a part of the template? I've got that in the pipeline already, so it's not going to take too long I hope.

tstuefe replied:
Fine for me

tstuefe reviewed (commented):
Almost good.

jdksjolen commented:
> > > The include guards should be placed before the includes.
> > 
> > 
> > Thanks!
> > > Would it make sense to put this in utilities/ instead of nmt/.
> > 
> > 
> > Do we have anyone else that wants to use this? The threshold in terms of API quality is typically higher for something in utilities/ than a local utility. I'm working on some improvements to this allocator already, maybe we can take a move to utilities/ together with those improvements?
> 
> Yes, I will use it to replace some code in Metaspace at least, but for that I need some more features (placing this thing atop of an existing memory range, and templatized index type).
> 
> Can all be done in a follow-up RFE.
> 
> > > Could we try another round of coming up with a better name for this utility? HomogenousObjectArray is eerily similar to G1's humongous object arrays. It's also not clear to me what makes this array an homogenous array. Is our other arrays non-homogenous?
> > 
> > 
> > I agree on the name part. We could call it a `HomogenousAllocator`, that does differentiate it meaningfully from something like `Arena` which can allocate anything. @tstuefe , arenas are also an example of the storage/lifetime of the objects from an allocator being bound to the lifetime of that allocator, so I think this makes sense.
> 
> Names are hard.
> 
> We already have two types of arenas in hotspots, plus glibc has arenas, so "arena" is not ideal either. Arena does usually not imply a free list either, nor homogenous sizes.
> 
> Thinking about it, Array already sort of implies homogeneous sizes, so maybe "homogeneous" is redundant.
> 
> ArrayWithFreeList? Unsexy but precise.

Agreed, let's go with `ArrayWithFreeList`.

> > Hi,
> > I've changed the data which is stored in `HOA` to be trivial and have added `static_assert`s to the methods. ... It's very important that the data we put into `HOA` is trivial as we do not call any destructors or copy constructors/operators on the code.
> 
> We don't, but why forbid the user from doing that? I may want to place a non-trivial object in there, e.g. one where I forbid copying altogether, or one with a single non-default ctor? What would be the problem with that?
> 
> Forbidding copying would be a good way to make sure the only instance exists at the place the array provides. So its address stable.

We can't have what you suggest at the moment, as objects aren't address stable without a fixed and non-growing memory area. Any form of constructor is actually fine, as those are called as they are.

tstuefe commented on src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:
```diff
@@ -40,64 +40,41 @@
   // Pick a prime number of buckets.
   // 4099 gives a 50% probability of collisions at 76 stacks (as per birthday problem).
   static const constexpr int default_table_size = 4099;
   int _table_size;
```
Pre-existing. Both table size and, arguably, the table itself are never changed, since we don't resize. At least size could be const, _table too if you move allocation to init list. (latter is a matter of taste, up to you).

tstuefe commented on src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:
```diff
@@ -120,15 +97,19 @@ class NativeCallStackStorage : public CHeapObj<mtNMT> {
+        _table[i] = TableEntryStorage::nil;
       }
     }
   }
```
Could you move this, and the dtor, over to the cpp file, too? You probably can remove the allocation.hpp include then. Less include deps are always good.

tstuefe commented on test/hotspot/gtest/nmt/test_arrayWithFreeList.cpp:
```diff
@@ -0,0 +1,152 @@
+    start = next_start;
+    return e;
+  }
+};
```
nit, add newline

tstuefe reviewed (approved):
I think this looks good now.

jdksjolen commented:
One last change: I've removed the `e char[sizeof(E)]` part of the union and made it `E e`. As the type is trivially copyable and trivially destructible this is fine. I realised that the alignment of `I` may be larger than the alignment of `E` if we're unlucky (storing a `char` for example) then that can lead to faulty behavior. We're now a bit safer.

afshin-zafari reviewed (approved):
Thanks, my comments are applied.

jdksjolen commented:
Thank you for the in-depth reviewing of this change.

/integrate

SoniaZaldana commented:
Hi @jdksjolen, 
I'm seeing some build failures after this PR got integrated. I filed [JDK-8335108](https://bugs.openjdk.org/browse/JDK-8335108) for reference.

jdksjolen commented:
> Hi @jdksjolen, I'm seeing some build failures after this PR got integrated. I filed [JDK-8335108](https://bugs.openjdk.org/browse/JDK-8335108) for reference.

Hi Sonia, thanks for contacting me. It seems like this is a C++20 issue, but Hotspot only supports up to C++14. Here's a Godbolt for reproducing the issue, change the `-std=c++14` to `-std=c++20` to see the issue: https://godbolt.org/z/eK5GaTceY

That ought to mean that I don't have to back out this PR and re-introduce it with the bug fixed, but I still think that we should fix the bug because we do want to get to C++20 some day :-).

There's something weird with the fact that you get builds with C++20 support enabled, not sure what your configure options are.

SoniaZaldana commented:
Hi @jdksjolen, thanks for looking into it! 

> Hi Sonia, thanks for contacting me. It seems like this is a C++20 issue, but Hotspot only supports up to C++14. Here's a Godbolt for reproducing the issue, change the `-std=c++14` to `-std=c++20` to see the issue: https://godbolt.org/z/eK5GaTceY
> 
> That ought to mean that I don't have to back out this PR and re-introduce it with the bug fixed, but I still think that we should fix the bug because we do want to get to C++20 some day :-).

Makes sense. I'll submit a follow-up PR to fix this. 

> There's something weird with the fact that you get builds with C++20 support enabled, not sure what your configure options are.

I'm not sure. I don't have any special configure options. I came across this error after a plain ```bash configure``` but I guess my installed g++ version is 14.1.1.

jdksjolen commented:
> _Mailing list message from [Thomas Stüfe](mailto:thomas.stuefe@gmail.com) on [hotspot-dev](mailto:hotspot-dev@mail.openjdk.org):_
> 
> Johan, I would just remove the noncopyable. Its questionable anyway, there is no reason one shouldnt be able to copy the AWFL, in particular because client code just uses indices, not pointers.
> 
> On Tue 25. Jun 2024 at 21:01, Sonia Zaldana Calles <szaldana at openjdk.org> wrote:
> 
> -------------- next part -------------- An HTML attachment was scrubbed... URL: <https://mail.openjdk.org/pipermail/hotspot-dev/attachments/20240625/f693178c/attachment.htm>

Whichever is fine by me. The class will still be non-copyable however, as GrowableArrayCHeap already is.
